client/lib/domains/utils.js

```diff
@@ -46,7 +46,7 @@ function getDomainNameFromReceiptOrCart( receipt, cart ) {
 		domainRegistration = cartItems.getDomainRegistrations( cart )[ 0 ];
 	}
 
-	if ( domainRegistration ) {
+	if ( domainRegistration && ( ! receipt || isEmpty( receipt.failed_purchases ) ) ) {
```

Comment:
I don't think this module should be concerned about the behavior of `checkout.jsx`. Even if it results in fewer lines of code overall, I think moving the checkout-specific conditions into checkout would be a good change (but I also might not fully understand what the other approach would look like, as I'm still a little confused as to why we need this).